mlflow/pyfunc/__init__.py

```diff
@@ -1648,12 +1662,158 @@ def _convert_struct_values(
+        -3:
+    ]
+
+    python_env = _get_python_env(Path(local_model_path))
+    env_sha = _get_virtualenv_name(python_env, local_model_path).split("-")[-1]
+    runtime_version, platform_machine = (
+        get_dbconnect_udf_sandbox_image_version_and_platform_machine(spark)
+    )
+
+    if prebuilt_env_sha != env_sha:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' does not match the model required environment."
+        )
+    if prebuilt_runtime_version != runtime_version:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' runtime version '{prebuilt_runtime_version}' "
+            f"does not match UDF sandbox runtime version {runtime_version}."
+        )
+    if prebuilt_runtime_version != runtime_version:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' platform machine '{prebuilt_platform_machine}' "
+            f"does not match UDF sandbox platform machine {prebuilt_platform_machine}."
+        )
+
+
+def _prebuild_env_internal(local_model_path, archive_name, save_path):
+    env_root_dir = os.path.join(_PREBUILD_ENV_ROOT_LOCATION, archive_name)
+    archive_path = os.path.join(save_path, archive_name + ".tar.gz")
+    if os.path.exists(env_root_dir):
+        shutil.rmtree(env_root_dir)
+    if os.path.exists(archive_path):
+        os.remove(archive_path)
+
     try:
-        from pyspark.sql.utils import is_remote
-    except ImportError:
-        return False
-    return is_remote()
+        pyfunc_backend = get_flavor_backend(
+            local_model_path,
+            env_manager="virtualenv",
+            install_mlflow=False,
+            create_env_root_dir=False,
+            env_root_dir=env_root_dir,
+        )
+
+        pyfunc_backend.prepare_env(model_uri=local_model_path, capture_output=False)
+        # exclude pip cache from the archive file.
+        shutil.rmtree(os.path.join(env_root_dir, "pip_cache_pkgs"))
+
+        return archive_directory(env_root_dir, archive_path)
+    finally:
+        shutil.rmtree(env_root_dir, ignore_errors=True)
+
+
+def prebuild_model_env(model_uri, save_path):
+    """
+    Prebuild model python environment and generate an archive file saved to provided
+    `save_path`.
+
+    Note:
+    You can only call `prebuild_model_env` in Databricks runtime,
+    and you can use the generated prebuilt env file in `mlflow.pyfunc.spark_udf` in
+    Databricks runtime or Databricks Connect client.
+    The generated model env file is named by:
+    'mlflow-{sha of python env config and dependencies}-{runtime version}-{platform machine}.tar.gz'
+    the file name shouldn't be modified,
+    and the prebuilt env can't be used across different Databricks runtime version or
+    different platform machine.
+    When you use the prebuilt env in other execution environment, MLflow will verify whether the
+    spark UDF sandbox environment matches the prebuilt env requirements.
+    """
```

Comment:
/ai remove "you" from docstring.

Reply:
The reason I'd avoid using "you" is the caller might not be a human. It could be a third-party library that depends on mlflow. Imagine transformers uses this function and I saw an error message like "You did ...". It sounds a bit strange because I didn't touch this function. Transformers did.